### ISSUE TITLE
Psycopg <class 'IndexError'> tuple index out of range

### ISSUE THREAD

**Issue by IlyaSukhanov** [aws/aws-xray-sdk-python#90]
```
  File "/var/task/sqlalchemy/engine/default.py", line 412, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/var/task/aws_xray_sdk/ext/psycopg2/patch.py", line 20, in _xray_traced_connect
    dbname = kwargs['dbname'] if 'dbname' in kwargs else re.search(r'dbname=(\S+)\b', args[0]).groups()[0]
```

**Comment by IlyaSukhanov** [comment]
I'm presuming the `host` `dbname` `port` `user` extraction logic does not support all the formats of connection strings that psycopg supports.

**Comment by IlyaSukhanov** [comment]
`psycopg2.connect()` Supports 3 variants of operation:
1. `psycopg2.connect("dbname=test user=postgres password=secret")`
2. `psycopg2.connect(dbname="test", user="postgres", password="secret")`
3. `psycopg2.connect()`

Variant 3 is ok if arguments are passed via env-vars:
> Any other connection parameter supported by the client library/server can be passed either in the connection string or as a keyword. The PostgreSQL documentation contains the complete list of the supported parameters. Also note that the same parameters can be passed to the client library using environment variables.

aws-xray-sdk-python does not support variant 3 or mix of variant 2 and 3.

**Comment by IlyaSukhanov** [comment]
It may not be obvious but use of patch_all() or patch of psycopg will cause any code error out and stack trace when opening connection to database. This is quite serious bug.

**Comment by IlyaSukhanov** [comment]
After some more digging around turns out problem was different but solution remained the same.

Psycopg supports passing of database name both as `dbname` and as `database`. SQL Alchemy uses the `database` variety triggering the bug.
https://github.com/psycopg/psycopg2/blob/master/lib/__init__.py#L96

There may still be a way to use method 3 as described in comment above but its not obvious. And psycopg will error out if there are no arguments on connect. https://github.com/psycopg/psycopg2/blob/master/lib/__init__.py#L122 Maybe if you set port via kwarg / dsn but the rest via env var?

**Comment by revmischa** [comment]
Getting this error when database is localhost

**Comment by Tankanow** [comment]
Is this still an issue or was this solved with 180de75?

**Comment by haotianw465** [comment]
@Tankanow this is fixed in the commit you mentioned. We will close this issue once the commit is released (which should be soon).

**Comment by haotianw465** [comment]
The fix is released in 2.2.0 https://pypi.org/project/aws-xray-sdk/. Closing.
